File: kindconnect/Driver_Service/src/main/java/com/kindconnect/Driver_Service/Service/DriverService.java
package com.kindconnect.Driver_Service.Service;

import java.util.List;

import org.springframework.stereotype.Service;

import com.kindconnect.Driver_Service.DTO.AvailablePickupDto;
import com.kindconnect.Driver_Service.DTO.DriverDeliveryDto;
import com.kindconnect.Driver_Service.Model.DonationClient;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class DriverService {

    private final DonationClient donationClient;
    private final DriverDashboardService dashboardService;

    // ================= STAGE 3 =================
    // View available donations
    public List<AvailablePickupDto> getAvailablePickups(String jwtToken) {
        return donationClient.getAvailablePickups(jwtToken);
    }

    // ================= STAGE 4 =================
    // Pickup donation
    public void pickupDonation(
            Long donationId,
            Long driverUserId,
            String jwtToken
    ) {

        // 1️⃣ Call Donation Service (state change)
        donationClient.pickupDonation(donationId, jwtToken);

        // 2️⃣ Record action (Stage 6)
        dashboardService.recordPickup(driverUserId, donationId);
    }

    // ================= STAGE 5 =================
    // Deliver donation
    public void deliverDonation(
            Long donationId,
            Long driverUserId,
            String jwtToken
    ) {

        // 1️⃣ Call Donation Service (state change)
        donationClient.deliverDonation(donationId, jwtToken);

        // 2️⃣ Record action (Stage 6)
        dashboardService.recordDelivery(driverUserId, donationId);
    }

    // ================= DRIVER DELIVERIES =================
    // Get driver's in-transit deliveries
    public List<DriverDeliveryDto> getInTransitDeliveries(String jwtToken) {
        return donationClient.getInTransitDeliveries(jwtToken);
    }

    // Get driver's completed deliveries
    public List<DriverDeliveryDto> getCompletedDeliveries(String jwtToken) {
        return donationClient.getCompletedDeliveries(jwtToken);
    }
}

File: kindconnect/frontend/src/services/driverService.js
import api from './api';

// ============= DRIVER PROFILE =============
export const createDriverProfile = async (profileData) => {
  const response = await api.post('/profiles/driver', profileData);
  return response.data;
};

export const getDriverProfile = async () => {
  const response = await api.get('/profiles/driver/me');
  return response.data;
};

export const updateDriverAvailability = async (available) => {
  const response = await api.put('/profiles/driver/availability', null, {
    params: { available }
  });
  return response.data;
};

// ============= DRIVER DASHBOARD =============
export const getDriverDashboard = async () => {
  const response = await api.get('/driver/dashboard');
  return response.data;
};

// ============= AVAILABLE PICKUPS =============
export const getAvailablePickups = async () => {
  const response = await api.get('/driver/donations/available');
  return response.data;
};

// ============= PICKUP DONATION =============
export const pickupDonation = async (donationId) => {
  const response = await api.put(`/driver/donations/${donationId}/pickup`);
  return response.data;
};

// ============= DELIVER DONATION =============
export const deliverDonation = async (donationId) => {
  const response = await api.put(`/driver/donations/${donationId}/deliver`);
  return response.data;
};

// ============= GET DONATION DETAILS =============
export const getDonationDetails = async (donationId) => {
  const response = await api.get(`/donations/${donationId}`);
  return response.data;
};

// ============= GET DRIVER DELIVERIES =============
export const getInTransitDeliveries = async () => {
  const response = await api.get('/donations/driver/in-transit');
  return response.data;
};

export const getCompletedDeliveries = async () => {
  const response = await api.get('/donations/driver/completed');
  return response.data;
};

export default {
  createDriverProfile,
  getDriverProfile,
  updateDriverAvailability,
  getDriverDashboard,
  getAvailablePickups,
  pickupDonation,
  deliverDonation,
  getDonationDetails,
  getInTransitDeliveries,
  getCompletedDeliveries
};
